### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by wojtke**

### CODE REVIEW

File: nets.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        self.out = torch.nn.Sequential()
+        channels = [
+            self.g_unet.out_channels if self.readout != self.readout_cat else self.g_unet.out_channels * 3,
+            *[mlp_channels for _ in range(mlp_layers-1)],
+            out_channels
+        ]
+        for i in range(mlp_layers):
+            self.out.add_module(f"mlp_dropout_{i}", torch.nn.Dropout(dropout) if dropout > 0 else torch.nn.Identity())
+            self.out.add_module(f"mlp_{i}", torch.nn.Linear(channels[i], channels[i + 1]))
+            self.out.add_module(f"mlp_act_{i}", torch.nn.ELU()) if i < mlp_layers - 1 else None
```
Comment (j-adamczyk) [inline nets.py]: Refactor into separate MLP class

File: nets.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+            out_channels (int): Number of output channels - number of classes.
+        """
+        super().__init__()
+        self.g_unet = g_unet
+        self.readout = {
+            "mean": torch_geometric.nn.global_mean_pool,
+            "max": torch_geometric.nn.global_max_pool,
+            "add": torch_geometric.nn.global_add_pool,
+            "cat": self.readout_cat,
+        }[readout.lower()]
```
Comment (j-adamczyk) [inline nets.py]: Change into 2 separate lines

File: nets.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    def __init__(self, g_unet: torch.nn.Module, readout: str, mlp_channels, mlp_layers, dropout, out_channels):
+        """Initializes the GNN.
+
+        Args:
+            g_unet (torch.nn.Module): Graph u-net module.
+            readout (str): Readout function. One of "add", "mean", "max", "cat".
+            mlp_channels (int): Number of channels in the MLP layers.
+            mlp_layers (int): Number of MLP layers.
+            dropout (float): Dropout probability for MLP layers.
+            out_channels (int): Number of output channels - number of classes.
```
Comment (j-adamczyk) [inline nets.py]: Move typing to `__init__` declaration

File: nets.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+from torch_geometric.utils import (
+    add_self_loops,
+    remove_self_loops,
+    sort_edge_index,
+)
+from torch_geometric.utils.repeat import repeat
+from torch_sparse import spspmm
+
+
+def create_model(in_channels, out_channels, **hyperparams):
```
Comment (j-adamczyk) [inline nets.py]: Change to GNN class, composing GraphUnet and MLP

Comment (j-adamczyk) [reply nets.py]: Also add typing

File: nets.py (hunk 1 above)
Comment (j-adamczyk) [inline nets.py]: Change to regular if

File: nets.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        }[readout.lower()]
+
+        self.out = torch.nn.Sequential()
+        channels = [
+            self.g_unet.out_channels if self.readout != self.readout_cat else self.g_unet.out_channels * 3,
+            *[mlp_channels for _ in range(mlp_layers-1)],
+            out_channels
+        ]
+        for i in range(mlp_layers):
+            self.out.add_module(f"mlp_dropout_{i}", torch.nn.Dropout(dropout) if dropout > 0 else torch.nn.Identity())
```
Comment (j-adamczyk) [inline nets.py]: Do not append Identity if there is no dropout

File: nets.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    )
+
+    gnn = GNN(
+        g_unet=g_unet,
+        readout=hyperparams["readout"],
+        mlp_channels=hyperparams["channels_classifier"],
+        mlp_layers=hyperparams["layers_classifier"],
+        dropout=hyperparams["dropout_classifier"],
+        out_channels=out_channels,
+    )
```
Comment (j-adamczyk) [inline nets.py]: Parse in a separate function, e.g. `parse_hyperparams`, returning a dataclass

File: data_utils.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    Args:
+        dataset_name (str): Name of the dataset to load.
+    """
+    dataset = TUDataset(root="data/TUDataset", name=dataset_name)
+
+    return dataset
+
+
+def get_dataset_cv_incicies(dataset, k_folds=10):
```
Comment (j-adamczyk) [inline data_utils.py]: Typo: "indices"

Comment (j-adamczyk) [reply data_utils.py]: Also add typing

File: model_selection.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        n_trials=n_trials,
+        timeout=timeout,
+        n_jobs=n_jobs,
+    )
+
+
+if __name__ == "__main__":
+    args = parse_args()
+    print_args(args)
+    main(**vars(args))
```
Comment (j-adamczyk) [inline model_selection.py]: Move to `main.py`. It should:
1. Arguments:
  - dataset name
  - convolution name
  - pooling name
2. Parse arguments:
  - task: "regression", "single_task_classification", "multi_task_classification"
  - evaluation metric: "MSE", "accuracy", "AUROC"
  - num_labels: int (1 for regression, number of classes for single task classif., number of labels for multi task classif.)
  - map convolution name to convolution class
  - map pooling name to pooling class
  - load splits from disk (lists of indices, list of 2-element tuples); if not present, throw error
3. Testing loop:
  - iterate through train/test splits
  - split train into train and valid
  - optimize model on train and valid
  - 3 times:
    - re-split, get new train/valid split
    - train new model with early stopping on new valid set
    - test new model, append test result to inner test results list
  - get avg test result from inner test results
  - append test result to outer test results
4. Report results (avg and stddev of outer test results)

`split_dataset.py`
argument: dataset name
splits dataset, saves indices on disk

splits:
  1. fold 1: train/test
  2. fold 2: train/test
  ...
  10. fold 10: train/test

Comment (j-adamczyk) [reply model_selection.py]: Implement evaluation functions for different tasks:
```
def evaluate_regression(y_true, y_pred)
```

```
def evaluate_single_task_classification(y_true, y_pred, metric: str):
    metric_fun = ... # map metric name to function
```

```
def evaluate_multi_task_classification(y_true, y_pred, metric: str):
    for each column in y_true and y_pred:
        col_metric = evaluate_single_task_classification(y_true[:, i], y_pred[:, i], metric)
    return avg(column metrics)
```

**Review by j-adamczyk — CHANGES_REQUESTED** [review]
